apps/aecore/src/aec_tx_pool_sync.erl

```diff
@@ -0,0 +1,477 @@
+    Fun = fun() ->
+              timer:sleep(Delay),
+              SerUnfolds = serialize_unfolds(Unfolds),
+              case aec_peer_connection:tx_pool_sync_unfold(PeerId, SerUnfolds) of
+                  {ok, NewSerUnfolds} ->
+                      {ok, deserialize_unfolds(NewSerUnfolds)};
+                  {error, <<"tree_not_ready">>} ->
+                      {ok, tree_not_ready, Unfolds, Delay};
+                  Err = {error, _} ->
+                      Err
+              end
+          end,
+    do_local_action(Sync, unfold, Fun, ?LOCAL_ACTION_TIMEOUT);
+do_local_action(Sync = #sync{ peer_id = PeerId }, {get, TxHashes}) ->
+    Fun = fun() -> aec_peer_connection:tx_pool_sync_get(PeerId, TxHashes) end,
+    do_local_action(Sync, get, Fun, ?LOCAL_ACTION_TIMEOUT);
+do_local_action(Sync = #sync{ peer_id = PeerId }, {finish, Done}) ->
+    %% Try to send finish to Peer - if it fails it fails but at least
+    %% we tried.
+    Fun = fun() ->
+              try
+                  aec_peer_connection:tx_pool_sync_finish(PeerId, Done)
+              catch _:_ ->
+                  ok
+              end,
+              {ok, Done}
+          end,
+    do_local_action(Sync, finish, Fun, ?LOCAL_ACTION_TIMEOUT).
+
+%% To avoid locking up main process (and deadlock in the interaction with
+%% aec_peer_connection) we do all requests in a worker process. The worker
+%% is linked, and we also start a timer when spawning it. The timeout is
+%% 10s which should be sufficently more than the request timeout (7s) in
+%% aec_peer_connection.
+do_local_action(#sync{ id = Ref }, Action, ActionFun, Timeout) ->
+    Self = self(),
+    Fun = fun() ->
+              %% Receive the timeout message ref so
+              %% we can cancel it on success...
+              receive {go, SendRef} -> ok end,
+              Res =
+                  try
+                      ActionFun()
+                  catch _:Reason ->
+                      {error, Reason}
+                  end,
+              Self ! {local_action, Ref, SendRef, Action, Res}
+          end,
+    Pid        = spawn_link(Fun),
+    TimeoutMsg = {local_action, Ref, Pid, Action},
+    SendRef    = erlang:start_timer(Timeout, self(), TimeoutMsg),
+    Pid ! {go, SendRef}.
+
+kill_local_action(Worker, Ref, Action) ->
+    erlang:exit(Worker, kill),
+    %% Flush any late arriving result
+    receive {local_action, Ref, _, Action, _} -> ok
+    after 0 -> ok end.
+
+do_get(TxHashes) ->
+    lists:concat([ do_get_(Tx) || Tx <- TxHashes ]).
+
+do_get_(TxHash) ->
+    try
+        STx = aec_db:get_signed_tx(TxHash),
+        [STx]
+    catch _:_ ->
+        []
+    end.
+
+unfold({node, Path, Node}, Tree) ->
+    aeu_mp_trees:unfold(Path, Node, Tree);
+unfold({leaf, Path}, _Tree) ->
+    {leaf, Path};
+unfold({subtree, Path}, Tree) ->
+    get_subtree(Path, Tree);
+unfold({key, Key}, _Tree) ->
+    [{key, Key}].
+
+get_subtree(Key, _Tree) when bit_size(Key) =:= 32 * 8 ->
```

Comment:
This clause looks an optimization for performing 0 lookups instead of two. Ok. (I would prefer not having the clause at all for maintainability though debatable.)

Reply:
It is actually necessary since the iterator on aeu_mp_trees will start from a Path *not including Path itself*.